// @justindfuller/verse-of-the-day/dist/utility/logging/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.$log = undefined;

var _appRootPath = require('app-root-path');

var _appRootPath2 = _interopRequireDefault(_appRootPath);

var _bunyan = require('bunyan');

var _bunyan2 = _interopRequireDefault(_bunyan);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function $log() {
  var environment = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : process.env.NODE_ENV;

  var streamOptions = {
    production: {
      level: 'warn',
      path: _appRootPath2.default + '/logs/server.log'
    },
    development: {
      level: 'trace',
      stream: process.stdout,
      src: true
    },
    test: {
      level: 'trace',
      stream: process.stdout
    }
  };

  return environment !== 'development' && environment ? _bunyan2.default.createLogger({
    name: 'server',
    streams: [streamOptions[environment]]
  }) : console;
}

exports.$log = $log;
exports.default = $log;
